test/pytest/receiver/test_receiver.py

```diff
@@ -0,0 +1,151 @@
+    plugin_handler.start()
+    plugin_handler.put((metadata, 1))
+    plugin_handler.stop(0)
+    mock_plugin.process.assert_not_called()
+    assert "ERROR" in caplog.text
+
+
+def test_plugin_fail_process(mock_plugin, caplog):
+    plugin_handler = PluginHandler("mock_plugin", {"foo": 1}, "bar", log)
+    mock_plugin.process.side_effect = Exception()
+    plugin_handler.start()
+    plugin_handler.put((metadata, 1))
+    plugin_handler.stop()
+    mock_plugin.stop.assert_called_once_with()
+    assert "ERROR" in caplog.text
+
+
+def test_plugin_fail_stop(mock_plugin, caplog):
+    plugin_handler = PluginHandler("mock_plugin", {"foo": 1}, "bar", log)
+    mock_plugin.stop.side_effect = Exception()
+    plugin_handler.start()
+    plugin_handler.put((metadata, 1))
+    plugin_handler.stop()
+    assert "ERROR" in caplog.text
+
+
+def test_plugin_slow_init(mock_plugin, caplog):
+    plugin_handler = PluginHandler("mock_plugin", {"foo": 1}, "bar", log)
+    mock_plugin_m.Plugin.side_effect = lambda *args, **kwargs: time.sleep(100)
+    t0 = time.time()
+    plugin_handler.start()
+    plugin_handler.put((metadata, 1))
+    plugin_handler.stop(0)
+    mock_plugin.process.assert_not_called()
+    assert time.time() - t0 < 10
+    assert "timeout" in caplog.text
+
+
+def test_plugin_slow_setup(mock_plugin, caplog):
+    plugin_handler = PluginHandler("mock_plugin", {"foo": 1}, "bar", log)
+    mock_plugin.setup.side_effect = lambda *args, **kwargs: time.sleep(100)
+    t0 = time.time()
+    plugin_handler.start()
+    plugin_handler.put((metadata, 1))
+    plugin_handler.stop(0)
+    mock_plugin.process.assert_not_called()
+    assert time.time() - t0 < 10
+    assert "timeout" in caplog.text
+
+
+def test_plugin_slow_process(mock_plugin, caplog):
+    plugin_handler = PluginHandler("mock_plugin", {"foo": 1}, "bar", log)
+    mock_plugin.process.side_effect = lambda *args, **kwargs: time.sleep(100)
+    t0 = time.time()
+    plugin_handler.start()
+    plugin_handler.put((metadata, 1))
+    plugin_handler.stop(0)
+    assert time.time() - t0 < 10
+    assert "timeout" in caplog.text
+
+
+def test_plugin_slow_stop(mock_plugin, caplog):
+    plugin_handler = PluginHandler("mock_plugin", {"foo": 1}, "bar", log)
+    mock_plugin.stop.side_effect = lambda *args, **kwargs: time.sleep(100)
+    t0 = time.time()
+    plugin_handler.start()
+    plugin_handler.put((metadata, 1))
+    plugin_handler.stop(0)
+    assert time.time() - t0 < 10
+    assert "timeout" in caplog.text
+
+
+def test_plugin_queue_overlow(mock_plugin, caplog):
+    plugin_handler = PluginHandler("mock_plugin", {"foo": 1}, "bar", log)
+    mock_plugin.process.side_effect = lambda *args, **kwargs: time.sleep(100)
+    t0 = time.time()
+    plugin_handler.start()
+    for i in range(10005):
+        plugin_handler.put((metadata, 1))
+    plugin_handler.stop(0)
+    assert time.time() - t0 < 10
+    assert plugin_handler.plugin_queue.qsize() == 10000
+    assert "Skipping" in caplog.text
```

Comment:
are the assertion tests of the time to check whether the plugin delay in a thread causes a delay outside it?

Reply:
Partly. When using a queue with an internal maximum size, `put` blocks when the queue is full. This tests asserts that the manual queue size limit is non-blocking. I added a docstring to clarify.